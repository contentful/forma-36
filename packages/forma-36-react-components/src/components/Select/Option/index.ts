export * from './Option';
export { default } from './Option';
