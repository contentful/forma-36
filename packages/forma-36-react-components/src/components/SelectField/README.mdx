---
title: 'Select field'
type: 'component'
status: 'stable'
slug: /components/select-field/
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/SelectField'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-selectfield--default'
---

Detailed documentation will be provided soon. For now, please have a look into the storybook for implementation details.

```jsx
<SelectField
  name="optionSelect"
  id="optionSelect"
  labelText="Label"
  selectProps="large"
  helpText="helpText"
>
  <Option value="optionOne">Option 1</Option>
  <Option value="optionTwo">Long Option 2</Option>
</SelectField>
```
