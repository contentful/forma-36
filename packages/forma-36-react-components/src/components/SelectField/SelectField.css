@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.SelectField {
  display: block;
}

.SelectField__label-wrapper {
  display: flex;
  font-size: var(--font-size-m);
  align-items: center;
}

.SelectField__label-link {
  margin-left: auto;
  margin-bottom: calc(1rem * (8 / var(--font-base-default)));
}
