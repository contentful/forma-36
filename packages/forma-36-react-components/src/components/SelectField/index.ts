export { SelectField } from './SelectField';
export type { SelectFieldProps } from './SelectField';
