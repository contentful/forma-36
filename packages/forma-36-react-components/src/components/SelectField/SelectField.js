import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import ValidationMessage from '../ValidationMessage';
import FormLabel from '../FormLabel';
import HelpText from '../HelpText';
import Select from '../Select/Select';
import TextLink from '../TextLink';
import styles from './SelectField.css';

class SelectField extends React.Component {
  static propTypes = {
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    validationMessage: PropTypes.string,
    testId: PropTypes.string,
    children: PropTypes.node.isRequired,
    extraClassNames: PropTypes.string,
    formLabelProps: PropTypes.object,
    textLinkProps: PropTypes.object,
    selectProps: PropTypes.object,
    name: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    labelText: PropTypes.string,
    helpText: PropTypes.string,
    required: PropTypes.bool,
    onChange: PropTypes.func,
    onBlur: PropTypes.func,
  };

  static defaultProps = {
    validationMessage: undefined,
    extraClassNames: undefined,
    testId: 'cf-ui-select-field',
    formLabelProps: undefined,
    textLinkProps: undefined,
    selectProps: undefined,
    labelText: undefined,
    onChange: undefined,
    helpText: undefined,
    onBlur: undefined,
    value: undefined,
    required: false,
  };

  state = { value: this.props.value };

  componentWillReceiveProps(nextProps) {
    if (this.props.value !== nextProps.value) {
      this.setState({
        value: nextProps.value,
      });
    }
  }

  // Store a copy of the value in state.
  // This is used by this component when the `countCharacters`
  // option is on
  handleOnChange = evt => {
    this.setState({ value: evt.target.value });
    if (this.props.onChange) {
      this.props.onChange(evt);
    }
  };

  render() {
    const {
      validationMessage,
      extraClassNames,
      children,
      selectProps,
      testId,
      formLabelProps,
      textLinkProps,
      labelText,
      helpText,
      required,
      onChange,
      onBlur,
      value,
      name,
      id,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.SelectField, extraClassNames);

    return (
      <div className={classNames} {...otherProps} data-test-id={testId}>
        <div className={styles['SelectField__label-wrapper']}>
          <FormLabel {...{ ...formLabelProps, htmlFor: id, required }}>
            {labelText}
          </FormLabel>
          {textLinkProps && (
            <TextLink
              {...{
                ...textLinkProps,
                extraClassNames: styles['SelectField__label-link'],
              }}
            >
              {textLinkProps.text}
            </TextLink>
          )}
        </div>
        <Select
          {...{
            hasError: !!validationMessage,
            name,
            id,
            onBlur,
            onChange: this.handleOnChange,
            value: this.state.value,
            required,
            ...selectProps,
          }}
        >
          {children}
        </Select>
        {validationMessage && (
          <ValidationMessage
            extraClassNames={styles['SelectField__validation-message']}
          >
            {validationMessage}
          </ValidationMessage>
        )}
        {helpText && (
          <div className={styles.SelectField__hints}>
            {helpText && (
              <HelpText extraClassNames={styles['SelectField__help-text']}>
                {helpText}
              </HelpText>
            )}
          </div>
        )}
      </div>
    );
  }
}

export default SelectField;
