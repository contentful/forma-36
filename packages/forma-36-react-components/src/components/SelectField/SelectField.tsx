import React, {
  useState,
  useEffect,
  ChangeEvent,
  FocusEventHandler,
  ChangeEventHandler,
} from 'react';
import cn from 'classnames';
import ValidationMessage from '../ValidationMessage';
import FormLabel, { FormLabelProps } from '../FormLabel';
import HelpText from '../HelpText';
import Select, { SelectProps } from '../Select';
import TextLink, { TextLinkProps } from '../TextLink';
import styles from './SelectField.css';

export interface SelectFieldProps {
  name: string;
  id: string;
  labelText: string;
  children: React.ReactNode;
  value?: string;
  validationMessage?: string;
  formLabelProps?: Partial<FormLabelProps>;
  textLinkProps?: Partial<TextLinkProps>;
  selectProps?: Partial<SelectProps>;
  helpText?: string;
  required?: boolean;
  onChange?: ChangeEventHandler;
  onBlur?: FocusEventHandler;
  testId?: string;
  className?: string;
}

export interface SelectFieldState {
  value?: string;
}

const defaultProps: Partial<SelectFieldProps> = {
  testId: 'cf-ui-select-field',
  required: false,
};

export const SelectField = (props: SelectFieldProps) => {
  const {
    validationMessage,
    className,
    children,
    selectProps,
    testId,
    formLabelProps,
    textLinkProps,
    labelText,
    helpText,
    required,
    onChange,
    onBlur,
    value,
    name,
    id,
    ...otherProps
  } = props;
  const [valueState, setValueState] = useState<string | undefined>(value);

  // Store a copy of the value in state.
  // This is used by this component when the `countCharacters`
  // option is on
  const handleOnChange = (e: ChangeEvent<HTMLSelectElement>) => {
    setValueState(e.currentTarget.value);
    if (onChange) {
      onChange(e);
    }
  };

  useEffect(() => {
    setValueState(value);
  }, [value]);

  const classNames = cn(styles['SelectField'], className);

  return (
    <div className={classNames} {...otherProps} data-test-id={testId}>
      <div className={styles['SelectField__label-wrapper']}>
        <FormLabel {...{ ...formLabelProps, htmlFor: id, required }}>
          {labelText}
        </FormLabel>
        {textLinkProps && (
          <TextLink
            {...{
              ...textLinkProps,
              className: styles['SelectField__label-link'],
            }}
          >
            {textLinkProps.text}
          </TextLink>
        )}
      </div>
      <Select
        {...{
          hasError: !!validationMessage,
          name,
          id,
          onBlur,
          onChange: handleOnChange,
          value: valueState,
          required,
          ...selectProps,
        }}
      >
        {children}
      </Select>
      {validationMessage && (
        <ValidationMessage
          className={styles['SelectField__validation-message']}
        >
          {validationMessage}
        </ValidationMessage>
      )}
      {helpText && (
        <div className={styles['SelectField__hints']}>
          {helpText && (
            <HelpText className={styles['SelectField__help-text']}>
              {helpText}
            </HelpText>
          )}
        </div>
      )}
    </div>
  );
};

SelectField.defaultProps = defaultProps;

export default SelectField;
