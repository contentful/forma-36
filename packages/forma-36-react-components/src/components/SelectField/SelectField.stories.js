import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import SelectField from './SelectField';
import Option from '../Select/Option';

storiesOf('Components|SelectField', module).add('default', () => (
  <SelectField
    extraClassNames={text('Extra Class Names', '')}
    required={boolean('Required', false)}
    formLabelProps={{
      requiredText: text('Required Text', undefined),
    }}
    name="optionSelect"
    id="optionSelect"
    labelText={text('Label', 'Label')}
    value={select(
      'Value',
      {
        'Option 1': 'optionOne',
        'Option 2': 'optionTwo',
      },
      'optionOne',
    )}
    onBlur={action('onBlur')}
    onChange={action('onChange')}
    selectProps={{
      isDisabled: boolean('Disabled', false),
      width: select(
        'Width',
        {
          'Full (default)': 'full',
          large: 'large',
          medium: 'medium',
          small: 'small',
        },
        'full',
      ),
    }}
    helpText={text('Help Text', '')}
    validationMessage={text('Validation Message', '')}
  >
    <Option value="optionOne">Option 1</Option>
    <Option value="optionTwo">Long Option 2</Option>
  </SelectField>
));
