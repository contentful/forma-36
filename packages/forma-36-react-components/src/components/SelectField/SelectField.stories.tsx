import React from 'react';
import SelectField, { SelectFieldProps } from './SelectField';
import Option from '../Select/Option';

export default {
  title: 'Form Elements/SelectField',
  component: SelectField,
  parameters: {
    propTypes: [SelectField['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
  },
};

export const Basic = (args: SelectFieldProps) => (
  <SelectField {...args}>
    <Option value="optionOne">Option 1</Option>
    <Option value="optionTwo">Long Option 2</Option>
  </SelectField>
);

Basic.args = {
  formLabelProps: {
    requiredText: 'Required Text',
  },
  name: 'optionSelect',
  labelText: 'labelText',
  value: 'optionOne',
  selectProps: {
    width: 'full',
  },
  helpText: 'helpText',
};
