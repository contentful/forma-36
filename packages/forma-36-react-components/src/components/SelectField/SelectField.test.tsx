import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import SelectField from './SelectField';
import Option from '../Select/Option';

it('renders the component', () => {
  const output = shallow(
    <SelectField id="optionSelect" name="optionSelect" labelText="Label">
      <Option value="optionOne">Option 1</Option>
    </SelectField>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component as disabled', () => {
  const output = shallow(
    <SelectField
      id="optionSelect"
      name="optionSelect"
      labelText="Label"
      selectProps={{
        isDisabled: true,
      }}
    >
      <Option value="optionOne">Option 1</Option>
    </SelectField>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component as required', () => {
  const output = shallow(
    <SelectField id="optionSelect" name="optionSelect" labelText="Label">
      <Option value="optionOne">Option 1</Option>
    </SelectField>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with a validation message', () => {
  const output = shallow(
    <SelectField
      id="optionSelect"
      name="optionSelect"
      validationMessage="not valid"
      labelText="Label"
    >
      <Option value="optionOne">Option 1</Option>
    </SelectField>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with a help text', () => {
  const output = shallow(
    <SelectField
      labelText="Label"
      id="optionSelect"
      name="optionSelect"
      helpText="please select an option"
    >
      <Option value="optionOne">Option 1</Option>
    </SelectField>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with a value', () => {
  const output = shallow(
    <SelectField
      id="optionSelect"
      name="optionSelect"
      value="optionOne"
      labelText="Label"
    >
      <Option value="optionOne">Option 1</Option>
    </SelectField>,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <SelectField id="optionSelect" name="optionSelect" labelText="Label">
      <Option value="optionOne">Option 1</Option>
    </SelectField>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
