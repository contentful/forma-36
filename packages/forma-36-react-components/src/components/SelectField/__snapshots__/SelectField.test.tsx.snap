// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="SelectField"
  data-test-id="cf-ui-select-field"
>
  <div
    class="SelectField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="optionSelect"
    >
      Label
    </label>
  </div>
  <div
    class="Select__wrapper Select--full"
  >
    <select
      aria-label="optionSelect"
      class="Select"
      data-test-id="cf-ui-select"
      id="optionSelect"
      name="optionSelect"
    >
      <option
        data-test-id="cf-ui-select-option"
        value="optionOne"
      >
        Option 1
      </option>
    </select>
    <svg
      class="Icon Icon--small Icon--muted Select__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </div>
</div>
`;

exports[`renders the component as disabled 1`] = `
<div
  class="SelectField"
  data-test-id="cf-ui-select-field"
>
  <div
    class="SelectField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="optionSelect"
    >
      Label
    </label>
  </div>
  <div
    class="Select__wrapper Select--full"
  >
    <select
      aria-label="optionSelect"
      class="Select Select--disabled"
      data-test-id="cf-ui-select"
      disabled=""
      id="optionSelect"
      name="optionSelect"
    >
      <option
        data-test-id="cf-ui-select-option"
        value="optionOne"
      >
        Option 1
      </option>
    </select>
    <svg
      class="Icon Icon--small Icon--muted Select__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </div>
</div>
`;

exports[`renders the component as required 1`] = `
<div
  class="SelectField"
  data-test-id="cf-ui-select-field"
>
  <div
    class="SelectField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="optionSelect"
    >
      Label
    </label>
  </div>
  <div
    class="Select__wrapper Select--full"
  >
    <select
      aria-label="optionSelect"
      class="Select"
      data-test-id="cf-ui-select"
      id="optionSelect"
      name="optionSelect"
    >
      <option
        data-test-id="cf-ui-select-option"
        value="optionOne"
      >
        Option 1
      </option>
    </select>
    <svg
      class="Icon Icon--small Icon--muted Select__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </div>
</div>
`;

exports[`renders the component with a help text 1`] = `
<div
  class="SelectField"
  data-test-id="cf-ui-select-field"
>
  <div
    class="SelectField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="optionSelect"
    >
      Label
    </label>
  </div>
  <div
    class="Select__wrapper Select--full"
  >
    <select
      aria-label="optionSelect"
      class="Select"
      data-test-id="cf-ui-select"
      id="optionSelect"
      name="optionSelect"
    >
      <option
        data-test-id="cf-ui-select-option"
        value="optionOne"
      >
        Option 1
      </option>
    </select>
    <svg
      class="Icon Icon--small Icon--muted Select__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </div>
  <div
    class="SelectField__hints"
  >
    <p
      class="HelpText SelectField__help-text"
      data-test-id="cf-ui-help-text"
    >
      please select an option
    </p>
  </div>
</div>
`;

exports[`renders the component with a validation message 1`] = `
<div
  class="SelectField"
  data-test-id="cf-ui-select-field"
>
  <div
    class="SelectField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="optionSelect"
    >
      Label
    </label>
  </div>
  <div
    class="Select__wrapper Select--full"
  >
    <select
      aria-label="optionSelect"
      class="Select Select--negative"
      data-test-id="cf-ui-select"
      id="optionSelect"
      name="optionSelect"
    >
      <option
        data-test-id="cf-ui-select-option"
        value="optionOne"
      >
        Option 1
      </option>
    </select>
    <svg
      class="Icon Icon--small Icon--muted Select__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </div>
  <div
    class="ValidationMessage SelectField__validation-message"
    data-test-id="cf-ui-validation-message"
  >
    <svg
      class="Icon Icon--small Icon--negative ValidationMessage__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
      />
    </svg>
    <p
      class="ValidationMessage__text"
    >
      not valid
    </p>
  </div>
</div>
`;

exports[`renders the component with a value 1`] = `
<div
  class="SelectField"
  data-test-id="cf-ui-select-field"
>
  <div
    class="SelectField__label-wrapper"
  >
    <label
      class="FormLabel"
      data-test-id="cf-ui-form-label"
      for="optionSelect"
    >
      Label
    </label>
  </div>
  <div
    class="Select__wrapper Select--full"
  >
    <select
      aria-label="optionSelect"
      class="Select"
      data-test-id="cf-ui-select"
      id="optionSelect"
      name="optionSelect"
    >
      <option
        data-test-id="cf-ui-select-option"
        value="optionOne"
      >
        Option 1
      </option>
    </select>
    <svg
      class="Icon Icon--small Icon--muted Select__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </div>
</div>
`;
